File: frontend/src/api/ingest.ts
import { apiPost, getApiBase, isApiConfigured } from './client'

export interface IngestMention {
  text: string
  entity_type: string
  resolved_entity_id?: string | null
  action: string
  source_excerpt?: string | null
}

export interface IngestResponse {
  status: string
  source_id: string
  source_type: string
  case_id?: string | null
  records_received: number
  entities_extracted: number
  entities_merged: number
  mentions: IngestMention[]
  extracted_text?: string | null
}

export interface PreviewEntity {
  text: string
  entity_type: string
  confidence: number
  source_excerpt?: string | null
}

export interface IngestPreviewResponse {
  engine: string
  spacy_available: boolean
  entities_extracted: number
  entities: PreviewEntity[]
}

export async function previewIngest(text: string): Promise<IngestPreviewResponse> {
  return apiPost<IngestPreviewResponse>('/ingest/preview', { text })
}

export async function ingestFirText(caseId: string, text: string): Promise<IngestResponse> {
  const body = new FormData()
  body.append('text', text)
  body.append('case_id', caseId)

  const res = await fetch(`${getApiBase()}/ingest/fir/text`, {
    method: 'POST',
    body,
  })
  if (!res.ok) {
    const msg = await res.text().catch(() => res.statusText)
    throw new Error(`Ingest failed: ${msg}`)
  }
  return res.json() as Promise<IngestResponse>
}

export async function ingestFirImage(caseId: string, file: File): Promise<IngestResponse> {
  const body = new FormData()
  body.append('file', file)
  body.append('case_id', caseId)

  const res = await fetch(`${getApiBase()}/ingest/fir/image`, {
    method: 'POST',
    body,
  })
  if (!res.ok) {
    const msg = await res.text().catch(() => res.statusText)
    throw new Error(`Image ingest failed: ${msg}`)
  }
  return res.json() as Promise<IngestResponse>
}

export function isIngestLive(): boolean {
  return isApiConfigured()
}

File: backend/app/schemas/ingest.py
from pydantic import BaseModel, Field


class ExtractedMention(BaseModel):
    text: str
    entity_type: str
    resolved_entity_id: str | None = None
    action: str
    source_excerpt: str | None = None
    match_reason: str | None = None
    requires_review: bool = False
    suggested_person_id: str | None = None


class IngestResponse(BaseModel):
    status: str
    source_id: str
    source_type: str
    case_id: str | None = None
    records_received: int = 0
    entities_extracted: int = 0
    entities_merged: int = 0
    mentions: list[ExtractedMention] = Field(default_factory=list)
    extracted_text: str | None = None


class PreviewEntity(BaseModel):
    text: str
    entity_type: str
    confidence: float
    source_excerpt: str | None = None


class IngestPreviewRequest(BaseModel):
    text: str


class IngestPreviewResponse(BaseModel):
    engine: str
    spacy_available: bool
    entities_extracted: int
    entities: list[PreviewEntity] = Field(default_factory=list)
    extracted_text: str | None = None
